### PR TITLE
Multimodality/sft bestfit sep classes

### PR DESCRIPTION
**Pull request by dtamayo-nlp**
# Motivation

The previous implementation worked correctly for SFT, but did not scale to pre-training workloads and eventually led to NCCL timeouts in tests with 4,000 files and billions of tokens.
This PR corrects issues raised in an older [PR](https://github.com/swiss-ai/Megatron-LM/pull/111).

## Features Introduced

The main bottleneck was the packing stage:

- bisect.insort() incurs O(N) insertion cost, making the overall complexity O(N log N + NxB), which degrades to ~O(N^2) for large corpora (N = number of documents, B = number of bins). Replacing it with SortedList reduced insertion to O(log N) and improved the complexity to O(N log N), but Python overhead remained a bottleneck at scale.
To fully address this, the default implementation was changed to C++ (python implementation is still available).

To correctly adapt the pipeline for pre-training, the following changes were also introduced:
- Support for mixed data regimes: the pipeline now handles both SFT (using `ApertusSFTDataset`) and raw pre-training data (using `GPTDataset`). SFT data must contain `apertus_sft` in the name, in a future PR, this will be cleaned up.
- Adaptation of the BFD algorithm to the `GPTDataset` class.
- Chunking instead of truncation in pre-training: documents longer than `seq_len` are split into fixed-length chunks prior to packing.
- `max_doc_per_bin` setting: for longer sequences, we want to avoid having too many small samples concatenated into a single sequence. This argument limits the number of documents packed during pre-training.

In the SFT case, we also incorporate the possibility of predicting **LONG CONTEXT TEXT** following this format for a specific long-context task:
```
<s>LONG CONTEXT TEXT<|system_start|><|system_end|>
<|developer_start|>Deliberation: disabled
Tool Capabilities: disabled<|developer_end|>
<|user_start|>How many times does the word "potato" appears?<|user_end|>
<|assistant_start|>The word potato appears 1234 times<|assistant_end|>
```

### Attention

As already discussed with @RaphaelKreft, if we have a very long document that is chunked into parts:

```text
<s> PART 1 | PART 2 | PART 3 </s>
```

Each part already fills the entire sample because its size equals `seq_len`, there is no need to add BOS/EOS tokens, given that there will be no additional documents.

### Loss Masking 1: Large User Inputs Could Cause Predicting User Queries

In the cooldown regime, we will not have these samples, but the current PR addresses the challenge of large user inputs by truncating the samples, as was previously implemented in the ApertusSFTDataset sampler.
IMO, we should pre-filter them to avoid wasting compute.

### Loss Masking 2: Loss Issues with Goldfish and Multimodality

This has been solved by separating the two dataset types (pre-training/SFT).

## Arguments Added

To launch SFT + pre-training, you will need to add these arguments to your launcher script:

```bash
--ap-sft
--ap-sft-mask-special-tokens
--ap-sft-pack-samples
--ap-sft-packing-strategy bfd
--pretraining-packing-strategy bfd
--max-docs-per-bin 64
```

### Test 1: Pre-training
Comparison between pre-training in `multimodality/main`, greedy pre-training in this branch, and BFD pre-training in this branch.

Conclusions:

- Greedy pre-training between `multimodality/main` and this branch is equivalent.
- Best-fit starts at a higher loss but quickly recovers.
- `max_doc_per_bin` in the best-fit approach helps significantly preserve throughput.

 
image
image

### Test 2: SFT
Comparison between SFT in `multimodality/sft-feat-bestfit-packing` and this branch. I trained both models on a retrieval task that asks the model to retrieve information from a dictionary. Given the C++ implementation, there are some differences after some samples, but the overall trend is similar.
image

### Test 3: Full Pre-training + SFT
Evaluating this new framework with pre-training + SFT data gives good results in both, short context and long context tasks.

### CODE REVIEW

File: megatron/core/datasets/blended_megatron_dataset_builder.py [hunk 1]
```diff
@@ -436,11 +437,17 @@ def _build_megatron_dataset_splits(
         Returns:
             List[Optional[MidLevelDataset]]: The MidLevelDataset (or None) per split
         """
+
+        if "apertus_sft" in dataset_path:
```
Comment (RaphaelKreft) [inline megatron/core/datasets/blended_megatron_dataset_builder.py]: We should do it in a more clever way. I would suggest:

Besides the dataset path and an optional blend ration, we add another optional arg for the dataset-type
For example for two data paths:

- Old: `--data-path 0.1 [PATH1] 0.9 [PATH2]`
- New Option 1: separate arg per dset path `--data-path **SFT** 0.1 [PATH1] **GPT** 0.9 [PATH2]`
- New Option 2:  use prefix in dset path`--data-path 0.1 SFT:[PATH1] 0.9 GPT:[PATH2]`

We can choose arbitrary type handles for **SFT** (Apertus SFT Dataset) or **GPT** (GPTDataset for pretrain data). Option 2 has the benefit of being easier to implement / less modifications necessary.

To modify:
- **For Option1: ** in megatron/core/datasets/utils.py - get_blend_from_list: add third branch for list of triples (dset-type, blend ratio and actual path)
- **For Option 2**: Add minimal parsing of datasetpath (split by : etc) in `megatron/core/datasets/blended_megatron_dataset_builder.py`

Comment (RaphaelKreft) [reply megatron/core/datasets/blended_megatron_dataset_builder.py]: of course we have to do document it in the ReadMe as well

File: megatron/training/datasets/bfd_pack.cpp
Comment (RaphaelKreft) [inline megatron/training/datasets/bfd_pack.cpp:1]: This file is a near duplicate of `megatron/core/datasets/bfd_pack.cpp`. Only real difference I can see is regarding how the zero length documents are treated and the max_docs_per_bin argument.

@dtamayo-nlp You think you could consolidate or is there another reason to have separate files here that I missed?

File: megatron/core/datasets/blended_megatron_dataset_builder.py (hunk 1 above)
Comment (RaphaelKreft) [reply megatron/core/datasets/blended_megatron_dataset_builder.py]: Another issue with thi is, that features from upstream are broken this way: MockGPTDataset, GPTFIMDataset, SFTDataset (JSONL) can no longer be selected through this builder.` --mock-data, --fim-data, --sft` are no-ops for class selection. Same in pretrain_gpt.py after the deletion of the dispatch block.

File: pretrain_gpt.py [hunk 2]
```diff
@@ -629,22 +639,10 @@ def train_valid_test_datasets_provider(train_val_test_num_samples, vp_stage=None
 
     config = core_gpt_dataset_config_from_args(args)
 
-    if args.ap_sft:
```
Comment (RaphaelKreft) [inline pretrain_gpt.py:632]: See comment on `blended_megatron_dataset_builder.py `regarding dataset selection logic

File: megatron/training/datasets/apertus_sft_dataset.py [hunk 3]
```diff
@@ -768,6 +869,17 @@ def _get_ltor_masks_and_position_ids(self, data: torch.Tensor, eos_indices: np.n
             if self.sft_plw_value > 0:
                 loss_mask[~assistant_mask] = self.sft_plw_value # value is 0 by default for full masking
 
+            # Also unmask tokens between BOS and <|system_start|> (pre-system content). Relevant for Long Context Tasks
```
Comment (RaphaelKreft) [inline megatron/training/datasets/apertus_sft_dataset.py:899]: for optimal performance this should be toggled via a flag as argument. Improves performance in case of non-long context data

**Review by RaphaelKreft — COMMENTED** [review]
Overall looks correct.

I added some comments to minor / medium issues that are mainly not related to correctness. Rather some style and broken upstream code.

Anyway, the initialize_sft_dataset script is broken due to new way datasets are distinguished by type (sft vs pretrain)

File: megatron/core/datasets/blended_megatron_dataset_builder.py (hunk 1 above)
Comment (dtamayo-nlp) [reply megatron/core/datasets/blended_megatron_dataset_builder.py]: Thanks for your contribution to this part, it's much cleaner now. I've merged your PR.

File: pretrain_gpt.py (hunk 2 above)
Comment (dtamayo-nlp) [reply pretrain_gpt.py:632]: I also resolve this comment given the merged PR: https://github.com/swiss-ai/Megatron-LM/pull/119

File: megatron/training/datasets/bfd_pack.cpp
Comment (dtamayo-nlp) [reply megatron/training/datasets/bfd_pack.cpp:1]: I've deduplicated the kernel in this [commit](https://github.com/swiss-ai/Megatron-LM/pull/113/changes/d12bc13130f9217e8c6a2492b670f1145a3358a5) and added the possible argument `--max-docs-per-bin-sft` in case the user wants to separate the maximum number of documents per bucket in a different way between pre-training/SFT.

File: megatron/training/datasets/apertus_sft_dataset.py (hunk 3 above)
Comment (dtamayo-nlp) [reply megatron/training/datasets/apertus_sft_dataset.py:899]: Added an argument `--ap-sft-long-ctx-loss` that controls the use of that masking loss in this [commit](https://github.com/swiss-ai/Megatron-LM/pull/113/changes/70d2b9a65e9ec3d197e3f8b4508dde75732924e6).

**Comment by RaphaelKreft** [comment]
From my side lgtm
